docs/src/manuals/performance-tips.md

```diff
@@ -0,0 +1,136 @@
+
+First of, a bad analogy: if RxInfer is the vessel taking you where you want to go, Julia is the ocean we're sailing. It is a good idea to have a basic understanding of this sea's many moods, so we'll start with an overview of the most essential Julia performance tips. For a deeper dive, see [Julia’s Performance Tips](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-tips) .
+
+
+
+---
+
+## General Julia Performance Tips
+
+**Summary:**
+- avoid global variables
+- write functions
+- ensure functions are type-stable
+- focus on steady-state performance
+
+### Avoid Untyped Global Variables
+**Why?**  
+In Julia, variables that are global (i.e., not inside a function or local scope) are **dynamically typed and slow**.  
+The compiler can't specialize code for unknown types, so it falls back to slower, general-purpose execution.  
+In RxInfer, this can seriously slow down:
+- **Model specifications** (building factor graphs)
+- **Non-linear function evaluations** (where fast numerical computation is essential)
+
+**Solution:**  
+Wrap your model-building code inside functions and ensure non-linear functions are also defined in a scoped, type-stable way. Making sure functions always return the same type of variable can boost performance.
+
+### Be Aware of Compilation Latency
+
+**Why?**  
+Julia uses **Just-In-Time (JIT)** compilation.  
+The **first time** you run a model (or even parts of it like `infer!`), Julia compiles the specialized machine code.  
+This can cause noticeable delays **only once**. Afterward, execution becomes extremely fast.
+
+**Tip:**  
+Don’t worry about long first-run times during development — focus on steady-state performance.
+
+---
+
+## Non-Linear Nodes
+
+### Prefer Linearization Over Unscented or CUI Methods
+
+**Why?**  
+- **Linearization** approximates non-linear functions by using **local linear models** around the current estimate.
+- **Unscented Transform** and **CUI (Central Unscented Integration)** are **more accurate** but involve **more function evaluations** and **matrix operations**, which are computationally heavier.
+
+Thus, **Linearization is much faster** and usually good enough unless you need extremely high precision.
+
+**Further Reading:**  
+See the “Deterministic Nodes” section of the RxInfer documentation for more on non-linear nodes.
+
+---
+
+## Free Energy Computation
+
+### Use `free_energy = Float64` Instead of `true`
+
+**Why?**  
+When you use `free_energy = true`, RxInfer needs to store more detailed internal states for diagnostic purposes.  
+When you specify `free_energy = Float64`, it reduces tracking to **minimal scalar computations**, making the process **leaner and faster**.
```

Comment:
The reason for `free_energy = Float64` is faster is because by default it stores the type of the computation as  `Real` in order to enable auto-differentiation